### ISSUE TITLE
Optional chaining with array access doesn't produce error for undefined

### ISSUE THREAD

**Issue by hipstersmoothie** [microsoft/TypeScript#35032]
**TypeScript Version:**  3.7.2 (also tested nightly)

**Search Terms:** optional-chaining

**Code**

```ts
interface Test {
  name: string;
}

const labels: Record<string, Test[] | undefined> = {
  foo: []
}

const key = 'bar'
const c = labels[key]?.[0]

c.name // it errors here
labels[key]?.[0].name // shouldn't this error?
```

**Expected behavior:**

It produces an error when trying to access a property on undefined in an optional chain

**Actual behavior:**

It allows the access, resulting in a run-time error

**Playground Link:** https://www.typescriptlang.org/play/?ts=3.8.0-dev.20191105&ssl=14&ssc=25&pln=1&pc=1#code/JYOwLgpgTgZghgYwgAgCoQM5mQbwFDLIhwC2EAXMllKAOYDceAvnnggPYhbIA2cARhB4ZKAJQgcoAEwA81OgBo0mMAG0AusgA+yAK4gpEGKAhSAfMgC8uAshjt2lDc1Ycu2ANYQAnleQByfjgof1Y+QWFVL291AH4AOlUABnV44jJkAHpMqgALdl0eKRB-bDBc4AxkaCh2KFjXTm4EP3ChDCifOMSUtjTSFGzkYGwauqrc6AggA

**Comment by jcalz** [comment]
This is working as intended due to [long short-circuiting](https://github.com/tc39/proposal-optional-chaining#long-short-circuiting).  After `?.`, subsequent chained property accesses/function calls/method calls happen only if the value before `?.` is non-nullish.   So `a?.b.c` is not equivalent to `(a?.b).c`.

**Comment by nmain** [comment]
> resulting in a run-time error

To be clear, `labels[key]?.[0].name` _does not_ result in a run-time error and this is why Typescript is correct.  Only the alternate form `const c = labels[key]?.[0]; c.name` results in a run-time error, but Typescript correctly flags this as a compile-time error as well.

**Comment by fatcerberus** [comment]
On a related but tangential note, optional chaining for arrays `?.[]` and function calls `?.()` is so ugly. 🙁  It always stops me in my tracks and takes me several seconds to figure out what it’s doing.  Ew.

(I know this isn’t TypeScript’s fault)

**Comment by hipstersmoothie** [comment]
> To be clear, labels[key]?.[0].name does not result in a run-time error and this is why Typescript is correct.

But I can see the error in the output?

in:

```tsx
labels[key]?.[0].name; // shouldn't this error?
```

out:

```js
(_a = labels[key]) === null || _a === void 0 ? void 0 : _a[0].name; // shouldn't this error?
```

I can clearly see that this will result in an error if the `foo` array is empty

The following works as I expect:

```
labels[key]?.[0]?.name;
(labels[key]?.[0]).name; // <- paren makes the error appear
(labels[key]?.[0])?.name;
```

Reading the spec I see under `Long short-circuiting`:

> Note that the check for nullity is made on `a` only. If, for example, `a` is not null, but `a.b` is null, a TypeError will be thrown when attempting to access the property "c" of `a.b`

It seems to be that TypeScript should tell me about this type error if it can. 

> So a?.b.c is not equivalent to (a?.b).c.

This to me seems like it could be large source of bugs. I get that short circuiting isn't supposed to run the RHS if the left is undefined, but it seems odd to me everything on the RHS defaults to defined.

EDIT: Playing around with It further maybe this is all just a misunderstanding by me on arrays in typescript. I was under the impression that [this](https://www.typescriptlang.org/play/?ts=3.8.0-dev.20191105&ssl=7&ssc=1&pln=7&pc=21#code/JYOwLgpgTgZghgYwgAgCoQM5mQbwFDLIhwC2EAXMllKAOYDceAvnnggPYhbKRaXpYA2gF1kAXmQjWvMIIAMwgHTEy45AHIY7duqA) would produce a typescript error, but it does not.

**Comment by nmain** [comment]
> I can clearly see that this will result in an error if the `foo` array is empty

But in your example you're not looking at `key == "foo"`, you're looking at `key == "bar"` so the optional chaining stops things.  For `key == "foo"`, as you discovered later, https://github.com/microsoft/TypeScript/issues/11122 is the relevant issue.
